Amazon SQS Connector in New Template (#379)
mulesoft/docs-connectors: modules/ROOT/pages/amazon/amazon-sqs-connector-examples.adoc

. The first flow sends a message with metadata. Then it gets the count of the messages in the queue to validate that the message was sent.
. The second flow receives the message and logs the message body.

image::amazon/amazon-sqs-message-flow.png[sqs-message-flow]

This example uses variables for some field values. You can either replace the variables with their values in the code, or you can provide the values for each variable in the `src/main/resources/mule-artifact.properties` file.

Before you try the example, access Anypoint Studio, and check whether the Mule Palette displays entries for Amazon SNS and Amazon SQS. If not, follow the instructions in xref:amazon/amazon-sqs-connector-studio.adoc[Amazon SQS Studio Configuration] to install both connectors.

== Create a Flow to Send a Message

Start the first flow by sending a message to the queue:

. Create a new Mule project in Anypoint Studio.
. In the Mule Palette, search for *HTTP*, and select the *Listener* operation:
+
image:amazon/amazon-sqs-select-listener.png[select-listener]
+
. Drag the *Listener* operation onto the canvas.

> . In the *Listener* configuration, click *+* to next to the *Connector configuration* field to add a global element.

Reviewer: ```suggestion
. In the *Listener* configuration, click *+* next to the *Connector configuration* field to add a global element.
```

. In the Mule Palette, search for *SQS* and select the *Send message* operation.
+
image::amazon/amazon-sqs-select-send.png[select-send-message]
+
. Drag the *Send message* operation onto the canvas, to the right of the *Transform Message* component.
. In the *Send message* configuration, click *+* next to the *Connector configuration* field to add a global element.
. Configure the global element as follows:
+
[%header%autowidth.spread]
|===
|Field |Value
|Name |Name used to reference the configuration
|Access Key|Alphanumeric text string that uniquely identifies the user who owns the account
|Secret Key |Key that plays the role of a password
|Region Endpoint | Queue region
|Default Global Queue URL | Default Amazon SQS queue URL
|===
+
For example:
+
image::amazon/amazon-sqs-send-global-config.png[send-global-config]
+
. Select the *Configuration XML* tab to view the corresponding XML:
+
[source,dataweave,linenums]
----
<sqs:config name="Amazon_SQS_Configuration"
  doc:name="Amazon SQS Configuration"
	defaultQueueUrl="${sqs.queueUrl}">
	<sqs:basic-connection accessKey="${sqs.accessKey}"
	secretKey="${sqs.secretKey}"
	region="us-east-1"/>
</sqs:config>
----
+

> . Configure the Following fields on the properties window:

Reviewer: ```suggestion
. Configure the following fields in the properties window:
```

+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name for the connector operation.
|Connector Configuration |Global configuration you just created
|Message |`payload`
|Queue url |Amazon SQS queue URL. If provided, the value of this field takes precedence over the value of the *Default Global Queue URL* field on the *Global Configuration Elements* window.
|===
+
For example:
+
image::amazon/amazon-sqs-send-message.png[send-message]

=== Add a Logger Component to Display the Response in the Mule Console

. In the Mule Palette, search for *Logger*.
. Drag the component onto the canvas, to the right of the *Send Message* component.
. Configure the following fields:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name for the logger, such as `Log Response`
|Message |`+++Sent Message: `#[payload]`+++`
|Level |INFO (Default)
|===
+
For example:
+
image::amazon/amazon-sqs-logger.png[logger]

=== Obtain the Number of Messages in the Queue

. In the Mule Palette, search for *Amazon SQS*.
. Select the *Get approximate number of messages* operation and drag it onto the canvas, to the right of the *Logger* component.
+

> . Configure the following fields on the properties window:

Reviewer: ```suggestion
. Configure the following fields in the properties window:
```

+
[%header%autowidth.spread]
|===
|Field |Value
|Queue url |Amazon SQS queue URL
|===
+
For example:
+
image::amazon/amazon-sqs-get-message-count.png[get-message-count]

=== Add a Logger to Display the Number in the Mule Console

. In the Mule Palette, search for *Logger*.
. Drag the component onto the canvas, to the right of the *Get approximate number of messages* component.
. Configure the following fields:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Enter a name for the logger, such as `Log Count`
|Message |`+++Approx. messages in queue: `#[payload]`+++`
|Level |INFO (Default)
|===
+
For example:
+
image::amazon/amazon-sqs-logger2.png[log-message-count]

== Create a Flow to Receive Messages

> Create another flow to receive messages and log them before deleting them from the queue.

Reviewer: ```suggestion
Create another flow to receive messages and log them before they are deleted from the queue.
```

. In the Mule Palette, search for *SQS* and select the *Receive messages* operation.
+
image::amazon/amazon-sqs-select-receive.png[select-receive-messages]
+
. Drag the *Receive messages* operation onto the canvas.
. Configure the following fields in the properties window:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name for the connector operation.

> |Connector Configuration |Global configuration you created previously for Amazon SQS Connector

Reviewer: ```suggestion
|Connector Configuration |Global configuration you created previously for Amazon SQS Connector.
```

Author: I am not sure about the period here. I will leave this one for Kim.

= Amazon SQS Examples - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

== Example

This example sends a message with metadata to an Amazon SQS queue and then receives it from the queue. This example is split into the two flows:

. The first flow sends a message with metadata. Then it gets the count of the messages in the queue to validate that the message was sent.
. The second flow receives the message and logs the message body.

> image::amazon/amazon-sqs-send-message-flow.png[Send Message Operation Flow]

Reviewer 2: Can you please update screenshot of this flow with the latest version of connector? Version of this connector is 5.4.1 and it has a nice icon now

> image::amazon/amazon-sqs-receive-message-flow.png[Receive and Delete Message Operation Flow]

Reviewer 2: Same here, please update this screenshot with connector's latest version 5.4.1

=== Add a Transform Message Component to Attach the Metadata

. In the Mule Palette, search for *Transform Message*.
+
image::amazon/amazon-sqs-select-transform.png[select-transform]
+
. Drag the *Transform Message* component onto the canvas, to the right of the *Listener*.
. In the *Transform Message* configuration, overlay the brackets in the *Output* section with this XML:
+
[source,dataweave,linenums]
----
{
	delaySeconds: 0,
	body: "Hello World",
	messageAttributes: {
		"AccountId": {
			"stringValue" : "000123456",
			"dataType" : "String.AccountId"
		} as Object {
			class: "org.mule.extension.sqs.api.model.MessageAttributeValue"
		},
		"NumberId": {
			"stringValue" : "230.000000000000000001",
			"dataType" : "Number"
		} as Object {
			class : "org.mule.extension.sqs.api.model.MessageAttributeValue"
		}
	} as Object {
		class: "java.util.HashMap"
	}
} as Object {
	class: "org.mule.extension.sqs.api.model.Message"
}
----
+
image::amazon/amazon-sqs-transform-message.png[transform-message]

=== Add and Configure the SQS Send Message Operation

. In the Mule Palette, search for *SQS* and select the *Send message* operation.
+

> image::amazon/amazon-sqs-select-send.png[select-send-message]

Reviewer 2: Same here, please update this screenshot with connector's latest version 5.4.1

. In the *Send message* configuration, click *+* next to the *Connector configuration* field to add a global element.
. Configure the global element as follows:
+
[%header%autowidth.spread]
|===
|Field |Value
|Name |Name used to reference the configuration.
|Access Key|Alphanumeric text string that uniquely identifies the user who owns the account.
|Secret Key |Key that plays the role of a password.
|Region endpoint | Queue region
|===
+
For example:
+
image::amazon/amazon-sqs-send-global-config.png[send-global-config]
+
. Select the *Configuration XML* tab to view the corresponding XML:
+
[source,dataweave,linenums]
----
<sqs:config name="Amazon_SQS_Configuration"
  doc:name="Amazon SQS Configuration"
	defaultQueueUrl="${sqs.queueUrl}">
	<sqs:basic-connection accessKey="${sqs.accessKey}"
	secretKey="${sqs.secretKey}"
	region="us-east-1"/>
</sqs:config>
----
+
. Configure the Following fields on the properties window:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Enter a name for the connector operation.
|Connector Configuration |Select the global configuration you just created.

> |Message |`#[payload]`

Reviewer 2: Please use payload instead of #[payload]

|Queue url |Enter the queue URL, if you did not enter a default queue URL in the global element.
|===
+
For example:
+

> image::amazon/amazon-sqs-send-message.png[send-message]

Reviewer 2: Same here, please update this screenshot with connector's latest version 5.4.1
- Also here please use **payload** in the send message operation instead of **#[payload]**

=== Add a Logger Component to Display the Response in the Mule Console

. In the Mule Palette, search for *Logger*.
. Drag the component onto the canvas, next to the *Send Message* component.
. Configure the following fields:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Enter a name for the logger, such as `Log Response`.
|Message |Sent Message: `#[payload]`
|Level |INFO (Default)
|===
+
For example:
+
image::amazon/amazon-sqs-logger.png[logger]

=== Obtain the Number of Messages in the Queue

. In the Mule Palette, search for *Amazon SQS*.
. Select the *Get approximate number of messages* operation and drag it onto the canvas, next to the *Logger* component.
+
. Configure the following fields on the properties window:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Enter a name for the connector operation.

> |Connector Configuration |Select the global configuration you created previously for Amazon SQS Connector.

Reviewer 2: This configuration is not the configuration which **Get Approximate number of messages** operation has. Only  field **queue url** should be there

|Message |`#[payload]`
|Queue url |Enter the queue URL.
|===
+
For example:
+

> image::amazon/amazon-sqs-get-message-count.png[get-message-count]

Reviewer 2: Also this image is for **Send message** not for  **Get Approximate number of messages**. Please update it with the latest version of connector

=== Add a Logger to Display the Number in the Mule Console

. In the Mule Palette, search for *Logger*.
. Drag the component onto the canvas, next to the *Get Count of Messages in queue* component.
. Configure the following fields:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Enter a name for the logger, such as `Log Count`.
|Message |Approx. messages in queue: `#[payload]`
|Level |INFO (Default)
|===
+
For example:
+
image::amazon/amazon-sqs-logger2.png[log-message-count]

== Create a Flow to Receive Messages

Now create another flow to receive messages and log them before deleting them from the queue.

. In the Mule Palette, search for *SQS* and select the *Receivemessages* operation.
+

> image::amazon/amazon-sqs-select-receive.png[select-send-message]

Reviewer 2: Same here, please update this screenshot with connector's latest version 5.4.1

+
. Drag the *Receivemessages* operation onto the canvas.
. Configure the following fields in the properties window:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Enter a name for the connector operation.
|Connector Configuration |Select the global configuration you created previously for Amazon SQS Connector.
|Number of Messages |10
|Queue url |
|===
+
NOTE: If provided, the value of the *Queue URL* field takes precedence over the *Global Element Properties Queue URL* field value.
+
For example:
+

> image::amazon/amazon-sqs-receive-message-config.png[select-send-message]

Reviewer 2: Same here, the small icon will change, please update this screenshot with connector's latest version 5.4.1
